pkg/sensors/program/coll.go

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Tetragon
+
+package program
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/tetragon/pkg/logger"
+)
+
+// The idea of LoadedCollection is to keep loaded programs and maps
+// from ebpf.Collection to be used later for TestLoad* tests to verify
+// loaded programs and maps.
+//
+// We can't just iterate all load programs and maps for 2 reasons:
+//   - we could race with others loading bpf programs
+//   - we get limited name, which we usually cross
+//
+// The process of loading LoadedCollection is following:
+//
+//   - ebpf collection is loaded
+//
+//     coll := NewCollectionWithOptions
+//
+//   - copy all programs/maps from ebpf collection
+//
+//     copyLoadedCollection(lc, coll)
+//
+//   - collection is removed and only 'used' programs and maps stay loaded
+//     so we filter out unused programs/maps with
+//
+//     load.LC = filterLoadedCollection(lc)
+//
+// This was we have only programs/maps realted to our test and with
+// original long names.
+//
```

Comment:
Thanks for writing this down!

One thing that got me thinking is the following. You say that:
> // We can't just iterate all load programs and maps for 2 reasons:
> //   - we could race with others loading bpf programs

Couldn't we race with other programs between copyLoadedCollection() and filterLoadedCollection()? If we only use this for testing, the race might be acceptable but I'm curious.

Reply:
as we discussed later the filtering is based on program/map ID, so it won't race